README
Name
    Text::Diff::HTML - XHTML format for Text::Diff::Unified

Synopsis
        use Text::Diff;

        my $diff = diff "file1.txt", "file2.txt", { STYLE => 'Text::Diff::HTML' };
        my $diff = diff \$string1,   \$string2,   { STYLE => 'Text::Diff::HTML' };
        my $diff = diff \*FH1,       \*FH2,       { STYLE => 'Text::Diff::HTML' };
        my $diff = diff \&reader1,   \&reader2,   { STYLE => 'Text::Diff::HTML' };
        my $diff = diff \@records1,  \@records2,  { STYLE => 'Text::Diff::HTML' };
        my $diff = diff \@records1,  "file.txt",  { STYLE => 'Text::Diff::HTML' };

Description
    This class subclasses Text::Diff::Unified, a formatting class provided
    by the Text::Diff module, to add XHTML markup to the unified diff
    format. For details on the interface of the "diff()" function, see the
    Text::Diff documentation.

    In the XHTML formatted by this module, the contents of the diff returned
    by "diff()" are wrapped in a "<div>" element, as is each hunk of the
    diff. Within each hunk, all content is properly HTML encoded using
    HTML::Entities, and the various sections of the diff are marked up with
    the appropriate XHTML elements. The elements used are as follows:

    *   "<div class="file">"

        This element contains the entire contents of the diff "file"
        returned by "diff()". All of the following elements are subsumed by
        this one.

        *   "<span class="fileheader">"

            The header section for the files being "diff"ed, usually
            something like:

              --- in.txt    Thu Sep  1 12:51:03 2005
              +++ out.txt   Thu Sep  1 12:52:12 2005

            This element immediately follows the opening "file" "<div>"
            element.

        *   "<div class="hunk">"

            This element contains a single diff "hunk". Each hunk may
            contain the following elements:

            *   "<span class="hunkheader">"

                Header for a diff hunk. The hunk header is usually something
                like:

                  @@ -1,5 +1,7 @@

                This element immediately follows the opening "hunk" "<div>"
                element.

            *   "<span class="ctx">"

                Context around the important part of a "diff" hunk. These
                are contents that have *not* changed between the files being
                "diff"ed.

            *   "<ins>"

                Inserted content, each line starting with "+".

            *   "<del>"

                Deleted content, each line starting with "-".

            *   "<span class="hunkfooter">"

                The footer section of a hunk; contains no contents.

        *   "<span class="filefooter">"

            The footer section of a file; contains no contents.

    You may do whatever you like with these elements and classes; I highly
    recommend that you style them using CSS. You'll find an example CSS file
    in the eg directory in the Text-Diff-HTML distribution. You will also
    likely want to wrap the output of your diff in its own element (a
    "<div>" will do) styled with "white-space: pre".

See Also
    Text::Diff
    Algorithm::Diff

Support
    This module is stored in an open repository at the following address:

    <https://svn.kineticode.com/Text-Diff-HTML/trunk/>

    Patches against Text::Diff::HTML are welcome. Please send bug reports to
    <bug-text-diff-html@rt.cpan.org>.

Author
    David E. Wheeler <david@justatheory.com>

Copyright and License
    Copyright (c) 2005-2011 David E. Wheeler. Some Rights Reserved.

    This module is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Text::Diff::HTML.

0.08  2018-08-21T03:21:47
      - Update project metadata: The Perl Shop is now project
        maintainer, distribution homepage is now on MetaCPAN, and the
        project is now explicitly licensed under the Artistic License 2.0
        (instead of "the same terms as Perl itself") with an appropriate
        LICENSE file.

0.07  2011-06-26T05:48:03
      - Moved to [GitHub](http://github.com/theory/text-diff-html/).
      - Switched to a static README.md, rather than a generated README.

0.06  2009-04-07T16:13:05
      - Fixed a pasto in the "Support" section of the documentation.
      - Changed the documention from recommending that the diff output be
        wrapped in <pre> tags to suggesting instead that it be wrapped in an
        element that uses the CSS "white-space: pre". Reported by Pete Gillin
        (RT #24301).

0.05  2008-05-04T02:11:09
      - Added the "configure_requires" and "recommends" parameters to
        Build.PL.
      - Added a link to the Subversion repository.

0.04  2005-09-06T16:33:51
      - Swiched from <span class="ins"> and <span class="del"> to <ins>
        and <del> elements. Why didn't I notice those before? Sorry for yet
        another change in the XHTML, but this is more semantic, and I there-
        fore don't expect to change it again. Suggested by Lucas Carlson.
      - Rejiggered sample CSS in "eg/diff.css" for new XHTML elements and
        shorter background color specs.

0.03  2005-09-05T18:40:29
      - Improved CSS examples in "eg/diff.css" by using more specific
        selectors.
      - Minor code and documentation reformatting.
      

0.02  2005-09-04T17:01:59
      - Added "See Also" section to documentation.
      - Added <div>s for the whole file and for each hunk. Suggested by
        Marshall Roch.
      - Fixed a number of annoying documentation typos and improved the
        documentation of each XHTML element.

0.01  2005-09-03T18:29:21
      - Initial public release.


Build.PL
use Module::Build;

my $build = Module::Build->new(
    module_name        => 'Text::Diff::HTML',
    license            => 'perl',
    create_makefile_pl => 'traditional',
    create_readme      => 1,
    configure_requires => { 'Module::Build' => '0.2701' },
    recommends         => { 'Test::Pod'     => '1.41'   },
    build_requires     => {
        'Test::More'    => '0.17',
        'Module::Build' => '0.2701',
    },
    requires           => {
        Text::Diff     => '0.11',
        HTML::Entities => '0',
    },
    meta_merge => {
        resources => {
            homepage   => 'http://search.cpan.org/dist/Text-Diff-HTML/',
            bugtracker => 'http://github.com/theory/text-diff-html/issues/',
            repository => 'http://github.com/theory/text-diff-html/',
        }
    },
);
$build->create_build_script;


lib/Text/Diff/HTML.pm
package Text::Diff::HTML;

use strict;
use vars qw(@ISA $VERSION);
use HTML::Entities;
use Text::Diff (); # Just to be safe.

$VERSION = '0.08';
@ISA = qw(Text::Diff::Unified);

sub file_header {
    return '<div class="file"><span class="fileheader">'
           . encode_entities(shift->SUPER::file_header(@_))
           . '</span>';
}

sub hunk_header {
    return '<div class="hunk"><span class="hunkheader">'
           . encode_entities(shift->SUPER::hunk_header(@_))
           . '</span>';
}

sub hunk_footer {
    return '<span class="hunkfooter">'
           . encode_entities(shift->SUPER::hunk_footer(@_))
           . '</span></div>';
}

sub file_footer {
    return '<span class="filefooter">'
           . encode_entities(shift->SUPER::file_footer(@_))
           . '</span></div>';
}

# Each of the items in $seqs is an array reference. The first one has the
# contents of the first file and the second has the contents of the second
# file, all broken into hunks. $ops is an array reference of array references,
# one corresponding to each of the hunks in the sequences.
#
# The contents of each op in $ops tell us what to do with each hunk. Each op
# can have up to four items:
#
# 0: The index of the relevant hunk in the first file sequence.
# 1: The index of the relevant hunk in the second file sequence.
# 2: The opcode for the hunk, either '+', '-', or ' '.
# 3: A flag; not sure what this is, doesn't seem to apply to unified diffs.
#
# So what we do is figure out which op we have and output the relevant span
# element if it is different from the last op. Then we select the hunk from
# second sequence (SEQ_B_IDX) if it's '+' and the first sequence (SEQ_A_IDX)
# otherwise, and then output the opcode and the hunk.

use constant OPCODE    => 2; # "-", " ", "+"
use constant SEQ_A_IDX => 0;
use constant SEQ_B_IDX => 1;

my %code_map = (
    '+' => [ 'ins'              => 'ins' ],
    '-' => [ 'del'              => 'del' ],
    ' ' => [ 'span class="ctx"' => 'span' ]
);

sub hunk {
    shift;
    my $seqs = [ shift, shift ];
    my $ops  = shift;
    return unless @$ops;

    # Start the span element for the first opcode.
    my $last = $ops->[0][ OPCODE ];
    my $hunk = qq{<$code_map{ $last }->[0]>};

    # Output each line of the hunk.
    while (my $op = shift @$ops) {
        my $opcode = $op->[OPCODE];
        my $elem   = $code_map{ $opcode } or next;

        # Close the last span and start a new one for a new opcode.
        if ($opcode ne $last) {
            $hunk .= "</$code_map{ $last }->[1]><$elem->[0]>";
            $last  = $opcode;
        }

        # Output the appropriate line.
        my $idx = $opcode ne '+' ? SEQ_A_IDX : SEQ_B_IDX;
        $hunk  .= encode_entities("$opcode $seqs->[$idx][$op->[$idx]]");
    }

    return $hunk . "</$code_map{ $last }->[1]>";
}

1;
__END__

##############################################################################

=head1 Name

Text::Diff::HTML - XHTML format for Text::Diff::Unified

=head1 Synopsis

    use Text::Diff;

    my $diff = diff "file1.txt", "file2.txt", { STYLE => 'Text::Diff::HTML' };
    my $diff = diff \$string1,   \$string2,   { STYLE => 'Text::Diff::HTML' };
    my $diff = diff \*FH1,       \*FH2,       { STYLE => 'Text::Diff::HTML' };
    my $diff = diff \&reader1,   \&reader2,   { STYLE => 'Text::Diff::HTML' };
    my $diff = diff \@records1,  \@records2,  { STYLE => 'Text::Diff::HTML' };
    my $diff = diff \@records1,  "file.txt",  { STYLE => 'Text::Diff::HTML' };

=head1 Description

This class subclasses Text::Diff::Unified, a formatting class provided by the
L<Text::Diff> module, to add XHTML markup to the unified diff format. For
details on the interface of the C<diff()> function, see the L<Text::Diff>
documentation.

In the XHTML formatted by this module, the contents of the diff returned by
C<diff()> are wrapped in a C<< <div> >> element, as is each hunk of the diff.
Within each hunk, all content is properly HTML encoded using
L<HTML::Entities>, and the various sections of the diff are marked up with the
appropriate XHTML elements. The elements used are as follows:

=over

=item * C<< <div class="file"> >>

This element contains the entire contents of the diff "file" returned by
C<diff()>. All of the following elements are subsumed by this one.

=over

=item * C<< <span class="fileheader"> >>

The header section for the files being C<diff>ed, usually something like:

  --- in.txt	Thu Sep  1 12:51:03 2005
  +++ out.txt	Thu Sep  1 12:52:12 2005

This element immediately follows the opening "file" C<< <div> >> element.

=item * C<< <div class="hunk"> >>

This element contains a single diff "hunk". Each hunk may contain the
following elements:

=over

=item * C<< <span class="hunkheader"> >>

Header for a diff hunk. The hunk header is usually something like:

  @@ -1,5 +1,7 @@

This element immediately follows the opening "hunk" C<< <div> >> element.

=item * C<< <span class="ctx"> >>

Context around the important part of a C<diff> hunk. These are contents that
have I<not> changed between the files being C<diff>ed.

=item * C<< <ins> >>

Inserted content, each line starting with C<+>.

=item * C<< <del> >>

Deleted content, each line starting with C<->.

=item * C<< <span class="hunkfooter"> >>

The footer section of a hunk; contains no contents.

=back

=item * C<< <span class="filefooter"> >>

The footer section of a file; contains no contents.

=back

=back

You may do whatever you like with these elements and classes; I highly
recommend that you style them using CSS. You'll find an example CSS file in
the F<eg> directory in the Text-Diff-HTML distribution. You will also likely
want to wrap the output of your diff in its own element (a C<< <div> >> will
do) styled with "white-space: pre".

=head1 See Also

=over

=item L<Text::Diff>

=item L<Algorithm::Diff>

=back

=head1 Support

This module is stored in an open L<GitHub
repository|http://github.com/theory/text-diff-html/>. Feel free to fork and
contribute!

Please file bug reports via L<GitHub
Issues|http://github.com/theory/text-diff-html/issues/> or by sending mail to
L<bug-Text-Diff-HTML@rt.cpan.org|mailto:bug-Text-Diff-HTML@rt.cpan.org>.

=head1 Author

David E. Wheeler <david@justatheory.com>

Currently maintained by the developers of The Perl Shop <tps@cpan.org>.

=head1 Copyright and License

Copyright (c) 2005-2011 David E. Wheeler. Some Rights Reserved.

This module is free software; you can redistribute it and/or modify it under the
terms of the Artistic License version 2.0.

=cut
